File: extension/generate-icons.mjs
// Generates solid-colour PNG icons for the Chrome extension using pure Node.js
// Run with: node generate-icons.mjs
import { mkdirSync, writeFileSync } from "fs";
import { createHash } from "crypto";
import zlib from "zlib";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

function u32be(n) {
  const b = Buffer.alloc(4);
  b.writeUInt32BE(n, 0);
  return b;
}

function crc32(buf) {
  // Simple CRC32 without external deps
  let crc = 0xffffffff;
  const table = (() => {
    const t = new Uint32Array(256);
    for (let i = 0; i < 256; i++) {
      let c = i;
      for (let j = 0; j < 8; j++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
      t[i] = c;
    }
    return t;
  })();
  for (const byte of buf) crc = table[(crc ^ byte) & 0xff] ^ (crc >>> 8);
  return (crc ^ 0xffffffff) >>> 0;
}

function pngChunk(tag, data) {
  const tagBuf = Buffer.from(tag);
  const crcInput = Buffer.concat([tagBuf, data]);
  return Buffer.concat([u32be(data.length), tagBuf, data, u32be(crc32(crcInput))]);
}

function makePng(size, r, g, b) {
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(size, 0); // width
  ihdr.writeUInt32BE(size, 4); // height
  ihdr[8] = 8;  // bit depth
  ihdr[9] = 2;  // colour type: RGB
  ihdr[10] = 0; ihdr[11] = 0; ihdr[12] = 0;

  // Each row: filter byte 0x00 + RGB pixels
  const row = Buffer.alloc(1 + size * 3);
  row[0] = 0;
  for (let x = 0; x < size; x++) {
    row[1 + x * 3] = r;
    row[2 + x * 3] = g;
    row[3 + x * 3] = b;
  }
  const raw = Buffer.concat(Array(size).fill(row));
  const idat = zlib.deflateSync(raw);

  return Buffer.concat([
    Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]), // PNG signature
    pngChunk("IHDR", ihdr),
    pngChunk("IDAT", idat),
    pngChunk("IEND", Buffer.alloc(0)),
  ]);
}

// Indigo: #6366f1
const [R, G, B] = [99, 102, 241];
const iconsDir = path.join(__dirname, "icons");
mkdirSync(iconsDir, { recursive: true });

for (const size of [16, 48, 128]) {
  const file = path.join(iconsDir, `icon${size}.png`);
  writeFileSync(file, makePng(size, R, G, B));
  console.log(`  Created ${file}`);
}
console.log("Done.");

File: backend/generate_icons.py
"""
Run this script once to generate placeholder PNG icons for the Chrome extension.
Usage: python generate_icons.py
"""
import os
import struct
import zlib


def _make_png(size: int, r: int, g: int, b: int) -> bytes:
    """Build a minimal PNG with a solid colour — no external deps required."""
    def u32(n: int) -> bytes:
        return struct.pack(">I", n)

    def chunk(tag: bytes, data: bytes) -> bytes:
        return u32(len(data)) + tag + data + u32(zlib.crc32(tag + data) & 0xFFFFFFFF)

    # IHDR
    ihdr = struct.pack(">IIBBBBB", size, size, 8, 2, 0, 0, 0)

    # Raw pixel rows: filter byte 0x00 + RGB triplets
    row = bytes([0]) + bytes([r, g, b] * size)
    raw = row * size
    idat = zlib.compress(raw, 9)

    return (
        b"\x89PNG\r\n\x1a\n"
        + chunk(b"IHDR", ihdr)
        + chunk(b"IDAT", idat)
        + chunk(b"IEND", b"")
    )


def main():
    out_dir = os.path.join(os.path.dirname(__file__), "..", "extension", "icons")
    os.makedirs(out_dir, exist_ok=True)

    # Indigo #6366f1 → r=99 g=102 b=241
    r, g, b = 99, 102, 241

    for size in (16, 48, 128):
        path = os.path.join(out_dir, f"icon{size}.png")
        with open(path, "wb") as f:
            f.write(_make_png(size, r, g, b))
        print(f"  Created {path}")

    # Update manifest to reference the icons
    import json
    manifest_path = os.path.join(os.path.dirname(__file__), "..", "extension", "manifest.json")
    with open(manifest_path) as f:
        manifest = json.load(f)

    manifest["icons"] = {
        "16": "icons/icon16.png",
        "48": "icons/icon48.png",
        "128": "icons/icon128.png",
    }
    manifest.setdefault("action", {})["default_icon"] = {
        "16": "icons/icon16.png",
        "48": "icons/icon48.png",
    }
    with open(manifest_path, "w") as f:
        json.dump(manifest, f, indent=2)
    print("  Updated manifest.json with icon paths")
    print("Done! Load /extension in chrome://extensions.")


if __name__ == "__main__":
    main()
